### DISCUSSION TITLE
Suggestion: official recommendation when having parallel installs (Arm/M1 macs)

### DISCUSSION THREAD

**Discussion by carlwr** [Homebrew/discussions#844] [Everyday usage]
For a while to come, most Arm/M1 mac users will see a need for two parallel Homebrew installations, `/usr/local/homebrew` for Rosetta-2/Intel and `/opt/homebrew` for Arm.

Having looked around quiet a bit, in the documentation, on the forum here, among filed Issues, and on external sites (Stack Exchange, blogs etc.), I do see a need for an official recommendation on how to handle this. There are a lot of great answers and information - what I see a need for is a recommendation based on community consensus, provided officially in the documentation, rather than just relying on individual users/community members having answered questions.

(Note: I'm considering doing the edits and submitting a pull request on the documentation once a recommended approach has been established. Related: my question [Regarding contributing to the documentation](https://github.com/Homebrew/discussions/discussions/841))

# What to endorse, in that case?

Two different approaches are circulating:

- A.) Have a separate Terminal.app with "Open using Rosetta" checked
- B.) Adding something like `alias ibrew='arch --x86_64 /usr/local/Homebrew/bin/brew'` to the shell startup script, and make sure `/opt/homebrew/bin` appears in `PATH` before `/usr/local/bin`

I cast my vore for the B.) option.

With B.), the user explicitly invokes either native or emulated `brew`, but when using installed programs, typically the user relies on path precedence, so that the native version will always be invoked if it exists, with the emulated version invoked as a fallback.

# Environment variables

Stuff just seems to work with the B.) approach above, but there will be some inconsistency with Homebrew's environement variables.

The installer suggests adding `eval $(/usr/local/bin/brew shellenv)` etc. to the shell startup script, effectively resulting in e.g.

```
export HOMEBREW_PREFIX="/usr/local";
export HOMEBREW_CELLAR="/usr/local/Cellar";
export HOMEBREW_REPOSITORY="/usr/local/Homebrew";
export PATH="/usr/local/bin:/usr/local/sbin${PATH+:$PATH}";
export MANPATH="/usr/local/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/usr/local/share/info:${INFOPATH:-}";
```

By including the lines

```
eval $(/usr/local/bin/brew shellenv)
eval $(/opt/homebrew/bin/brew shellenv)
```

in the shell startup script the `PATH` will contain both paths and with the right precedence, but the `HOMEBREW_*` variables will be overwritten by the second line and hence point to `/opt/homebrew`.

If Homebrew or third party tools rely on these environment variables existing and being correct, approach B.) doesn't really do it.

However, Homebrew doesn't seem to use them, instead determining the repository from the location of the invoked `brew` binary. Illustration:
```
$ export HOMEBREW_PREFIX="/opt/homebrew" && arch --x86_64 /usr/local/Homebrew/bin/brew config | grep HOMEBREW_PREFIX
HOMEBREW_PREFIX: /usr/local
```

(In my view, this behaviour is great, and offers an elegant solution for handling parallel Homebrew installs - if it can be trusted.)

So:
1. Is there any risk of trouble with the actual install path being inconsistent with the `HOMEBREW_*`'s?
2. With parallel Homebrew installations, is the user still recommended to add the lines 
    ```
    eval $(/usr/local/bin/brew shellenv)
    eval $(/opt/homebrew/bin/brew shellenv)
    ```
    to the shell startup script, in that order?

**Comment by SMillerDev** [comment]
> Two different approaches are circulating:
    A.) Have a separate Terminal.app with "Open using Rosetta" checked
    B.) Adding something like alias ibrew='arch --x86_64 /usr/local/Homebrew/bin/brew' to the shell startup script, and make sure /opt/homebrew/bin appears in PATH before /usr/local/bin

They're both the same option (run one install in rosetta). How people enable Rosetta shouldn't be up to Homebrew if you ask me.

---

> Is there any risk of trouble with the actual install path being inconsistent with the HOMEBREW_*'s?

@Homebrew/brew might know best

> With parallel Homebrew installations, is the user still recommended to add the lines
```
    eval $(/usr/local/bin/brew shellenv)
    eval $(/opt/homebrew/bin/brew shellenv)
```
> to the shell startup script, in that order?

As far as i know there should be only one of these, if you need both you'll need to filter for yourself based on a rosetta check.

**Reply by carlwr** [reply]
> > Two different approaches are circulating:
> > A.) Have a separate Terminal.app with "Open using Rosetta" checked
> > B.) Adding something like alias ibrew='arch --x86_64 /usr/local/Homebrew/bin/brew' to the shell startup script, and make sure /opt/homebrew/bin appears in PATH before /usr/local/bin
> 
> They're both the same option (run one install in rosetta). How people enable Rosetta shouldn't be up to Homebrew if you ask me.

Installation wise they are the same.

Usage wise, using A.), the user would need some way to make the shell startup script update `PATH` with `/usr/local/homebrew` for the Rosetta-enabled Terminal.app, and `/opt/homebrew` for the native Terminal.app.

Further, for Homebrew-installed programs, the B.) option has the slight elegance of automatically choosing the Arm version if available, otherwise falling back to invoking the Intel version.

I'm open to being convinced that all this is beyond what Homebrew should give recommendations for. However, the installer currently does instruct the user to add `eval $((...)/bin/brew shellenv)` to the shell startup script - something that will currently not "just work" with either the A.) or B.) option.

> > With parallel Homebrew installations, is the user still recommended to add the lines
> 
> ```
>     eval $(/usr/local/bin/brew shellenv)
>     eval $(/opt/homebrew/bin/brew shellenv)
> ```
> 
> > to the shell startup script, in that order?
> 
> As far as i know there should be only one of these, if you need both you'll need to filter for yourself based on a rosetta check.

(Just a note: such a filter is possible assuming the A.) approach, and would also make that approach work.)

**Comment by carlocab** [comment]
`eval $(/usr/local/bin/brew shellenv)` should never be needed, as the `/usr/local` prefix takes away the need to make any additions/modifications to your environment variables. It'll also make `/usr/local/bin` appear twice in your `PATH`. (This doesn't hurt, but I know some people find this annoying.)

**Reply by carlwr** [reply]
> `eval $(/usr/local/bin/brew shellenv)` should never be needed, as the `/usr/local` prefix takes away the need to make any additions/modifications to your environment variables.

Do you mean that with an alias like `ibrew='arch --x86_64 /usr/local/Homebrew/bin/brew'`, there's no need to include `eval $(/usr/local/bin/brew shellenv)` in the shell startup script - in that case `eval $(/opt/homebrew/bin/brew shellenv)` is enough?

If that's what you meant, well, it's not necessary for executing the `brew` binary; the alias takes care of that; but it's still necessary to runt the programs installed as intel/Rosetta programs.

That is, unless you run them by invoking them with the full explicit path. That's still an option. But that requires the user to always prefix those invokations with `arch --x86_64`, or by always invoking them from a Rosetta-enabled Terminal.app.

> It'll also make `/usr/local/bin` appear twice in your `PATH`. (This doesn't hurt, but I know some people find this annoying.)

Oh - but this must always have been the case for Intel macs? (`/opt/homebrew shellenv` won't add `/usr/local/bin` to `PATH`, but `/usr/local/bin/brew shellenv` will - so I assume you mean that the double entries in `PATH` come from both this command and some other system script adding it.)

**Reply by SMillerDev** [reply]
> Oh - but this must always have been the case for Intel macs? (/opt/homebrew shellenv won't add /usr/local/bin to PATH, but /usr/local/bin/brew shellenv will - so I assume you mean that the double entries in PATH come from both this command and some other system script adding it.)

Apple already added `/usr/local` to the path. That was the reason brew installed there in the first place.
That's also the reason you don't need to run the shellenv command and the reason all tools can generally find brew programs without configuration.

**Reply by carlocab** [reply]
No; when I say there is no need, I mean that there is no need. You don't need an alias, nor do you need to explicitly add it to your `PATH`. It's already there. You also don't need to touch `MANPATH`, etc.

**Reply by carlwr** [reply]
I messed things up, sorry, yeah it makes sense now.

I had also missed that the installer's suggestion to add `eval $((...)/bin/brew shellenv)` to the shell startup script is only given when installing for the Arm architecture; I thought this had always been the case, also when installing to `/usr/local` - but as you say, that isn't needed and has never been.

**So, in summary**

When the user installs Homebrew on an Arm mac, the installer will suggest adding `eval $(/opt/homebrew/bin/brew shellenv)` to the shell startup script. If there is an earlier `/usr/local/homebrew` installation, installed programs there will be invoked if none is found under `/opt/homebrew` because `/usr/local/bin` is also in `PATH`, added by the system.

When running `brew` commands, the user is responsible for making sure:
* that it is executed under the desired architecture, i.e. Arm or Intel, e.g. by prefixing with `arch --x86_64` or using a Rosetta 2 terminal), and also
* that the desired `brew` (from either `/usr/local/homebrew` or `/opt/homebrew`) is called, either by trusting the path precedence or by invoking with a full path.

I think this has settled most of my concerns.

Still, the `HOMEBREW_*` environment variables will typically point to `/opt/homebrew` even when `brew` in `/usr/local/homebrew` is invoked, if the user doesn't take further action to avoid this, however, this doesn't seem to cause any problems.

**Reply by carlocab** [reply]
>that it is executed under the desired architecture, i.e. Arm or Intel, e.g. by prefixing with `arch --x86_64` or using a Rosetta 2 terminal), and also

I don't think this is needed either, actually. If you have Rosetta 2 installed it should just kick in when necessary without any intervention. So the only complication is making sure you call the `brew` installation you want, and this depends on which one comes first in your `PATH`.

**Comment by Rylan12** [comment]
Forgive me if I'm duplicating information that's already been said.

> * Is there any risk of trouble with the actual install path being inconsistent with the `HOMEBREW_*`'s?

Variables like `HOMEBREW_REPOSITORY` and `HOMEBREW_PREFIX` when used _internally_ are determined based on the location of `bin/brew` (I believe I'm correct in saying that these are `/usr/local/bin/brew` or `/opt/homebrew/bin/brew` by default). There is no way (that I know of) to override these using environment variables.

So I believe the answer to your question is no.

---

**Extra info**

You don't even need to set `HOMEBREW_PREFIX` or the like to run Homebrew successfully. On my (Intel) machine I don't have these variables set and have no issues.

Where you could potentially have a problem is with any formulae or custom/third-party scripts.

Formulae, to my knowledge, don't use dynamically determined paths for anything that might need a reference to one of these variables. They are almost exclusively hardcoded at install-time with the only potential exception being anything in a `postinstall` block or in the `test` block. However, even in these cases, we are using the _internal_ definitions rather than potentially incorrect environment variables. Of course, we have no control over third-party formulae so if you have any of those there is a potential for problems, but I doubt there will be any in most cases.

Where you're most likely to run into issues is with custom/third-party scripts/commands. You'll have to check these yourself and make modifications if needed. I doubt there will be any problems, though, you just may have to make some more aliases or modify those scripts to handle the differing locations.

---

Edit: I missed your comment here that pretty much sums it up:

> Still, the `HOMEBREW_*` environment variables will typically point to `/opt/homebrew` even when `brew` in `/usr/local/homebrew` is invoked, if the user doesn't take further action to avoid this, however, this doesn't seem to cause any problems.
